op-interop-mon/monitor/updater.go

```diff
@@ -0,0 +1,172 @@
+package monitor
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	supervisortypes "github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+var ErrLogNotFound = errors.New("log not found")
+
+// TODO: make this configurable
+var updateInterval = 1 * time.Second
+
+type UpdaterClient interface {
+	BlockReceipts(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) ([]*types.Receipt, error)
+}
+
+var _ UpdaterClient = &ethclient.Client{}
+
+// Updaters are responsible for updating jobs from a chain for the Maintainer to track
+type Updater interface {
+	Start(ctx context.Context) error
+	Enqueue(job *Job)
+	Stop() error
+}
+
+// RPCFinder connects to an Ethereum chain and extracts receipts in order to create jobs
+type RPCUpdater struct {
+	client  UpdaterClient
+	chainID eth.ChainID
+
+	// the duration after the terminal state is set that the job is considered expired
+	expireTime time.Duration
+
+	inbox           chan *Job
+	enqueueCallback func(*Job)
+	expireCallback  func(*Job)
+	closed          chan struct{}
+
+	log log.Logger
+}
+
+func NewUpdater(chainID eth.ChainID, client UpdaterClient, enqueueCallback func(*Job), expireCallback func(*Job), log log.Logger) *RPCUpdater {
+	return &RPCUpdater{
+		chainID:         chainID,
+		client:          client,
+		log:             log.New("component", "rpc_updater", "chain_id", chainID),
+		inbox:           make(chan *Job, 10_000),
+		enqueueCallback: enqueueCallback,
+		expireCallback:  expireCallback,
+		closed:          make(chan struct{}),
+		expireTime:      1 * time.Minute,
+	}
+}
+
+func (t *RPCUpdater) Start(ctx context.Context) error {
+	go t.Run(ctx)
+	return nil
+}
+
+func (t *RPCUpdater) Run(ctx context.Context) {
+	for {
+		select {
+		// if the finder is closed, close the inbox and outbox and end the loop
+		case <-t.closed:
+			t.log.Info("updater closed")
+			close(t.inbox)
+			return
+		// if the inbox has a new job, process the job and send the jobs to the outbox
+		case job := <-t.inbox:
+			err := t.UpdateJob(job)
+			if err != nil {
+				t.log.Error("error updating job", "error", err)
+				continue
+			}
+			if t.ShouldExpire(job) {
+				t.log.Info("job expired", "job", job.String())
+				t.expireCallback(job)
+				continue
+			}
+			t.enqueueCallback(job)
```

Comment:
It's a little hard to see how subsequent updates are triggered (say I didn't find the initiating message the first time, and I need to check again). 

How about we just put a ticker into the updater, and add a `case` to this select statement? Then the updater is in control of continually updating all jobs until it decides (or is signalled) to stop tracking a job. it wouldn't need to be driven other than spawning the jobs in the first place.

Reply:
Yeah, subsequent updates are managed by the maintainer, who passes the job back to the updater when it is time. The updaters work on a pure inbox/outbox system, and track items long term. They just update and return the job.

We *could* totally make the updaters hold jobs similar to how the Maintainer does. I like this suggestion, thanks :)